core/src/main/scala/org/apache/spark/deploy/RRunner.scala

```diff
@@ -100,15 +100,17 @@ object RRunner {
         builder.redirectErrorStream(true) // Ugly but needed for stdout and stderr to synchronize
         val process = builder.start()
 
-        new RedirectThread(process.getInputStream, System.out, "redirect R output").start()
+        val stdoutBuffer = new CircularBuffer(1024)
+        val output = new TeeOutputStream(System.out, stdoutBuffer)
+        new RedirectThread(process.getInputStream, output, "redirect R output").start()
 
-        process.waitFor()
+        val returnCode = process.waitFor()
+        if (returnCode != 0) {
+          throw SparkUserAppException(returnCode, Option(stdoutBuffer.toString))
+        }
       } finally {
         sparkRBackend.close()
       }
-      if (returnCode != 0) {
```

Comment:
@jeremyjliu, can you show before/after error messages? Seems like we redirect stderr. Doesn't that work?